runtime/allocatable.h

```diff
@@ -0,0 +1,88 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+// Defines APIs for Fortran runtime library support of code generated
+// to manipulate and query allocatable variables, dummy arguments, & components.
+#ifndef FORTRAN_RUNTIME_ALLOCATABLE_H_
+#define FORTRAN_RUNTIME_ALLOCATABLE_H_
+#include "descriptor.h"
+#include "entry-names.h"
+
+namespace Fortran::runtime {
+extern "C" {
+
+// Initializes the descriptor for an allocatable of intrinsic or derived type.
+// The incoming descriptor is treated as (and can be) uninitialized garbage.
+// Must be called for each allocatable variable as its scope comes into being.
+// The storage for the allocatable's descriptor must have already been
+// allocated to a size sufficient for the rank, corank, and type.
+// A descriptor must be initialized before being used for any purpose,
+// but needs reinitialization in a deallocated state only when there is
+// a change of type, rank, or corank.
+void RTNAME(AllocatableInitIntrinsic)(
+    Descriptor &, TypeCategory, int kind, int rank = 0, int corank = 0);
+void RTNAME(AllocatableInitCharacter)(Descriptor &, SubscriptValue length,
+    int kind = 1, int rank = 0, int corank = 0);
+void RTNAME(AllocatableInitDerived)(
+    Descriptor &, const DerivedType &, int rank = 0, int corank = 0);
+
+// For MOLD= and SOURCE= allocation with no (other) prior initialization.
+void RTNAME(AllocatableInitMold)(Descriptor &, const Descriptor &);
+
+// Explicitly sets the bounds and length type parameters of an initialized
+// deallocated allocatable.  (Not for use with MOLD= or SOURCE= in an
+// ALLOCATE statement; see below).  Length type parameters are indexed in
+// declaration order; i.e., 0 is the first length type parameter in the
+// deepest base type.
+void RTNAME(AllocatableSetBounds)(
+    Descriptor &, int dim, SubscriptValue lower, SubscriptValue upper);
+void RTNAME(AllocatableSetDerivedLength)(
+    Descriptor &, int which, SubscriptValue);
+
+// The upper bound is ignored for the last codimension.
+void RTNAME(AllocatableSetCoBounds)(
+    Descriptor &, int codim, SubscriptValue lower, SubscriptValue upper = 0);
+
+// Allocates an allocatable.  The allocatable descriptor must have been
+// initialized and its bounds and length type parameters set and must be
+// in a deallocated state.  For SOURCE= allocation of an uninitialized
+// descriptor, use AllocatableInitMold() before the first allocation.
+// On failure, if hasStat is true, returns a nonzero error code for
+// STAT= and (if present) fills in errMsg; if hasStat is false, the
+// image is terminated.  On success, leaves errMsg alone and returns zero.
+// Successfully allocated memory is initialized if the allocatable has a
+// derived type, and is always initialized by AllocatableAllocateSource().
+int RTNAME(AllocatableAllocate)(
+    Descriptor &, bool hasStat = false, Descriptor *errMsg = nullptr);
```

Comment:
Did you intend to leave the check that the allocatable is not already allocated to this function ?
The issue I see with this is that we may have already mutated the descriptor (to set the bounds, or type parameters...) and it case we realize now that it is already allocated and `hasStat` is true, we would have to restore the original information as per 9.7.6 point 5:
> “each allocate-object that was not successfully allocated or deallocated shall retain its previous allocation status or pointer association status

In general, though it might be safer to backup the descriptor when `STAT` is present in order to fulfill 9.7.6 p6. But if not, care may have to be taken here.

Reply:
Maybe the APIs for setting bounds, length parameters, &c. should be capable of failing in the case of an allocated descriptor.  Or have a distinct API for checking for allocation for use in cases with STAT=; then the generated code can jump around the mutations.  Backing up a descriptor seems complicated to me.